src/Symfony/Component/Console/Command/Command.php

```diff
@@ -303,6 +303,27 @@ public function run(InputInterface $input, OutputInterface $output): int
      */
     public function complete(CompletionInput $input, CompletionSuggestions $suggestions): void
     {
+        $definition = $this->getDefinition();
+        $values = null;
+        if (CompletionInput::TYPE_OPTION_VALUE === $input->getCompletionType() && $definition->hasOption($input->getCompletionName())) {
+            $values = $definition->getOption($input->getCompletionName())->getValues();
+        } elseif (CompletionInput::TYPE_ARGUMENT_VALUE === $input->getCompletionType() && $definition->hasArgument($input->getCompletionName())) {
+            $values = $definition->getArgument($input->getCompletionName())->getValues();
+        }
+        if (null === $values) {
+            return;
+        }
+        if ($values instanceof \Closure) {
+            $values = $values($input);
+            if (null === $values) {
+                return;
+            }
+            if (!is_iterable($values)) {
+                throw new LogicException(sprintf('Callable for %s %s must return an iterable or null. Got "%s".', $input->getCompletionType(), $input->getCompletionName(), get_debug_type($values)));
+            }
+        }
+
+        $suggestions->suggestValues(is_array($values) ? $values : iterator_to_array($values));
```

Comment:
All this code could be moved to `CompleteCommand`, but that would break the `CommandCompletionTester`.